### ISSUE TITLE
"Show Context" doesn't show surrounding lines when multiple lines have the same timestamp

### ISSUE THREAD

**Issue by critchtionary** [grafana/grafana#64966]
**What happened**:
We're using Loki to aggregrate logs for our Jenkins builds which are scraped by FluentBit. It's not uncommon for multiple lines to have the same timestamp to the millisecond (which is the precision they are included in Jenkins logs) e.g. stack traces. 

When viewing these logs in the Explore mode in Grafana, it seems that intended order of the lines are preserved.  For example, I created a Jenkins job that wrote 100 lines to a file (prefixed with a random number to show the Explore mode wasn't just sorting them) and ran `cat` on the file. When querying the logs in Grafana, they appear in the order I would expect:

![image](https://user-images.githubusercontent.com/12222103/225940957-de716b2b-84c7-4a8c-a540-86228d035fbf.png)

However, if I filter for a specific line e.g. `|= "This is line 20"` and then click "Show context", the surrounding lines are not the same as what is displayed when not filtering.

![image](https://user-images.githubusercontent.com/12222103/225941376-d4f23bf2-da0a-4ef2-b1e3-0957a2f2dee6.png)

**What you expected to happen**:
"Show context" should show the same surrounding lines as when the log is not filtered.

**How to reproduce it (as minimally and precisely as possible)**:
Filter on a line that has the same timestamp as other lines and click "Show Context"

**Anything else we need to know?**:

**Environment**:
- Grafana version: 9.4.3
- Data source type & version: Loki 2.6.1
- OS Grafana is installed on: Official Docker image
- User OS & Browser: macOS 12.6 & Firefox 111.0
- Grafana plugins: None
- Others:

**Comment by usmangt** [comment]
Hi @critchtionary,

Thanks for opening this issue.

Can you please provide your Loki and Fluent bit configuration file so that we can start working on this issue?

**Comment by critchtionary** [comment]
I've just checked and this issue does also occur for logs that are sent to Loki by Promtail (with the same timestamp as initially described), so I've attached Loki and Promtail configs. These have been generated from the Tanka installation.

[configs.zip](https://github.com/grafana/grafana/files/11016508/configs.zip)

**Comment by critchtionary** [comment]
I don't believe it is a duplicate, as my example all of the lines have the same labels and the issue only occurs when multiple lines have the same timestamp.

**Comment by svennergr** [comment]
> I don't believe it is a duplicate, as my example all of the lines have the same labels and the issue only occurs when multiple lines have the same timestamp.

Oh, I see. That's something discussed in https://github.com/grafana/grafana/issues/56071. Ordering of logs with the exact same timestamp can not be guaranteed. If you need the same ordering when ingested, you might want to enable the `increment_duplicate_timestamp` setting, which will autoincrement duplicate timestamps by nanosecond (https://grafana.com/docs/loki/latest/configuration/#limits_config).

**Comment by critchtionary** [comment]
> > I don't believe it is a duplicate, as my example all of the lines have the same labels and the issue only occurs when multiple lines have the same timestamp.
> 
> Oh, I see. That's something discussed in #56071. Ordering of logs with the exact same timestamp can not be guaranteed. If you need the same ordering when ingested, you might want to enable the `increment_duplicate_timestamp` setting, which will autoincrement duplicate timestamps by nanosecond (https://grafana.com/docs/loki/latest/configuration/#limits_config).

I didn't know about that config option, but it should fix this problem. Thanks for your help!

**Comment by critchtionary** [comment]
Unfortunately setting `increment_duplicate_timestamp` actually made viewing the log in Grafana worse. Not sure whether it's because of out-of-order writes (as mentioned in the config docs) or because we have multiple distributor replicas. Regardless it won't work for us.

![image](https://user-images.githubusercontent.com/12222103/227913229-a8bb752d-65f8-4cc7-9958-bde7a9ef79c3.png)

>  Ordering of logs with the exact same timestamp can not be guaranteed

Is there a reason why it always seems to be "correct" when browsing in Grafana and "incorrect" when viewing context?
